job-runner/chain/reaper.go

```diff
@@ -0,0 +1,630 @@
+
+		// Copy job data to every child job, even if it's not ready to be run yet.
+		// When a job has multiple parent jobs, it'll get job data copied from each
+		// parent, not just the last one to finish. Be careful - it's possible for
+		// parents to overwrite each other's job data if they set the same field.
+		for k, v := range job.Data {
+			nextJob.Data[k] = v
+		}
+		r.chainRepo.Set(r.chain)
+
+		// Enqueue any runnable child jobs.
+		if !r.chain.IsRunnable(nextJob.Id) {
+			nextJLogger.Infof("next job is not runnable - not enqueuing it")
+			continue
+		}
+		nextJLogger.Infof("next job is runnable - enqueuing it")
+		if r.chain.IsSequenceStartJob(nextJob.Id) {
+			// Starting a sequence, so increment sequence try count.
+			r.chain.IncrementSequenceTries(nextJob.Id)
+			seqLogger := r.logger.WithFields(log.Fields{"sequence_id": job.SequenceId, "sequence_try": r.chain.SequenceTries(job.Id)})
+			seqLogger.Info("starting try of sequence")
+		}
+		r.runJobChan <- nextJob
+	}
+}
+
+// Finalize determines the final state of the chain and sends it to the Request
+// Manager.
+func (r *RunningChainReaper) Finalize(complete bool) {
+	if complete {
+		r.logger.Infof("chain is done, all jobs finished successfully")
+		r.chain.SetState(proto.STATE_COMPLETE)
+	} else {
+		r.logger.Warn("chain is done, some jobs failed")
+		r.chain.SetState(proto.STATE_FAIL)
+	}
+	r.chainRepo.Set(r.chain)
+	r.sendFinalState()
+	r.chainRepo.Remove(r.chain.RequestId())
+}
+
+// -------------------------------------------------------------------------- //
+
+// Job Reaper for chains that are being suspended (stopped to be resumed later).
+type SuspendedChainReaper struct {
+	reaper
+	runnerRepo runner.Repo
+}
+
+// Run reaps jobs when they finish running. For each job reaped, if it's...
+// - completed: prepare subsequent jobs (copy jobData).
+// - failed:    prepare a sequence retry.
+// - stopped:   do nothing (job will be retried when chain is resumed).
+func (r *SuspendedChainReaper) Run() {
+	defer close(r.doneChan)
+
+	// Sleep a short time to prevent a race condition on the runner Repo. If a job
+	// was sent to runJobChan just before the chain was suspended, its runner might
+	// not have been created + added to the Repo yet. Wait so we can be sure all
+	// running jobs have runners in the Repo, so our checks to runnerRepo.Count
+	// will accurately reflect whether there are any running jobs left.
+	time.Sleep(runnerRepoWait)
+
+	// If there are already no jobs left to reap, the running reaper must have
+	// finished and finalized the chain before it got switched out for this reaper.
+	// There's nothing left to do, so return right away.
+	if r.runnerRepo.Count() == 0 {
+		return
+	}
+
+	// Reap jobs until there are no jobs left running, or the reaper is stopped.
+REAPER:
+	for r.runnerRepo.Count() > 0 {
+		select {
+		case job := <-r.doneJobChan:
+			r.Reap(job)
+		case <-time.After(runnerRepoWait):
```

Comment:
Why do we need this case? Can't we just select on doneJobChan and stopChan? Or is there a case where runnerRepo.Count() will be greater than 0, but we'll never receive anything on doneJobChan?

Reply:
Yeah, we need this case - it is possible for runnerRepo.Count to be > 0 even though there's no jobs left running (and so nothing will get sent to doneJobChan). The runner gets removed from the repo right after the job gets sent to doneJobChan in traverser.runJobs(). Reaping a job is pretty quick for suspended/stopped chains, so it's possible for the reaper to receive the job, reap it, and go back to check the `r.runnerRepo.Count() > 0` loop condition before traverser.runJobs() gets a chance to remove the runner from the repo. If that happens, we need this case or we'd be stuck in the loop.

I'll add a comment to that ^ effect.